server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -130,7 +121,9 @@ public ValuesSourceType getValuesSourceType() {
 
     @Override
     public boolean equals(Object o) {
-        if (!super.equals(o)) return false;
+        if (o instanceof MappedFieldType == false) {
```

Comment:
I've always been a fan of `if (o == null || getClass() != o.getClass())` for this. Every `equals` implementation is ugly, but at least mine bails early if the subtypes don't line up.